### PR TITLE
feat: Allow parallel HTTP requests

### PR DESCRIPTION
**Pull request by simster7**
Closes https://github.com/argoproj/argo-workflows/issues/6994

Signed-off-by: Simon Behar <simbeh7@gmail.com>

Don't bother creating a PR until you've done this:

* [ ] Run `make pre-commit -B` to fix codegen, lint, and commit message problems.

Create your PR as a draft.

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it
  does not need to pass.
* Once required tests have passed, you can make it "Ready for review".
* Say how how you tested your changes. If you changed the UI, attach screenshots.

Tips:

* If changes were requested, and you've made them, then dismiss the review to get it looked at again.
* Add you organization to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md) if you like.
* You can ask for help!

### CODE REVIEW

File: workflow/executor/agent.go [hunk 1]
```diff
@@ -8,6 +8,7 @@ import (
 	"net/http"
 	"os"
 
+	"github.com/argoproj/pkg/sync"
```
Comment (simster7) [inline workflow/executor/agent.go:11]: WIP implementation: not exhaustive or thorough, but feedback welcome

File: workflow/executor/agent.go [hunk 2]
```diff
@@ -31,79 +31,135 @@ type AgentExecutor struct {
...
 	defer runtimeutil.HandleCrash(runtimeutil.PanicHandlers...)
 
+	log.Info("Starting Agent")
+
+	agentTaskQueue := make(chan agentTask)
+	patchResponseQueue := make(chan patchResponse)
 	taskSetInterface := ae.WorkflowInterface.ArgoprojV1alpha1().WorkflowTaskSets(ae.Namespace)
+
+	go ae.patchWorker(ctx, taskSetInterface, patchResponseQueue)
+	for i := 0; i < 8; i++ {
```
Comment (simster7) [inline workflow/executor/agent.go:68]: Default 16? Configurable?

**Comment by simster7** [comment]
@alexec @sarabala1979 Could I get a review here?

**Comment by sarabala1979** [comment]
Sure I will do it

**Review by alexec — COMMENTED** [review]
One thought I have is that we probably want to limit the rate of patches:

1. A loop to watch for events, and put each task onto a work queue.
2. N workers (configurable) that execute that work and put the results onto a result queue.
3. A single worker that both (a) reads results from the result queue and builds the patch (b) periodically (say every 5s, configurable) sends the patch.

Challenges here:

* We loose atomicity, results are not persisted before the next HTTP request made. We need to make sure that when the pod shutdown, and in flight results are persisted.
* We can configure the concurrency at a workflow level, but not at a template level.

**Comment by simster7** [comment]
Thanks for the great feedback, I'll patch in some fixes

> we probably want to limit the rate of patches

Agreed.

> We need to make sure that when the pod shutdown, and in flight results are persisted.

Great catch.

> We can configure the concurrency at a workflow level, but not at a template level.

We can add multiple rate-limiters per template to the code easily for template rate-limit overrides, along with a workflow-level default rate-limit.

**Comment by alexec** [comment]
> We can add multiple rate-limiters per template to the code easily for template rate-limit overrides, along with a workflow-level default rate-limit.

I think rate-limiting is not something needed today, but we might want to keep the door open to it in the future .

**Comment by alexec** [comment]
Great work BTW.

File: workflow/executor/agent.go (hunk 2 above)
Comment (sarabala1979) [reply workflow/executor/agent.go:68]: configurable is good. can we do it through Env and the controller will set ENV when creating POD?

File: workflow/executor/agent.go [hunk 3]
```diff
@@ -31,79 +31,135 @@ type AgentExecutor struct {
...
+	}
+}
 
-					obj, err = taskSetInterface.Patch(ctx, ae.WorkflowName, types.MergePatchType, patch, metav1.PatchOptions{})
+func (ae *AgentExecutor) patchWorker(ctx context.Context, taskSetInterface v1alpha1.WorkflowTaskSetInterface, patchResponseQueue chan patchResponse) {
+	for patch := range patchResponseQueue {
+		log.WithFields(log.Fields{"workflow": ae.WorkflowName, "nodeID": patch.NodeId}).Error("Processing Patch")
 
-					log.WithField("taskset", obj).Infof("updated content, %s", patch)
+		obj, err := taskSetInterface.Patch(ctx, ae.WorkflowName, types.MergePatchType, patch.Patch, metav1.PatchOptions{})
```
Comment (sarabala1979) [inline workflow/executor/agent.go:132]: Can we add retry logic for transient errors?

**Comment by simster7** [comment]
@alexec @sarabala1979 Please take another look

**Comment by alexec** [comment]
I leave Bala to review this. He is expert.

File: workflow/executor/agent.go [hunk 4]
```diff
@@ -31,79 +33,165 @@ type AgentExecutor struct {
...
+			}
 
-					ae.CompleteTask[nodeID] = struct{}{}
+			log.WithFields(log.Fields{"workflow": ae.WorkflowName}).Info("Processing Patch")
 
-					if err != nil {
-						log.WithError(err).WithField("taskset", obj).Errorf("failed to update the taskset")
+			obj, err := taskSetInterface.Patch(ctx, ae.WorkflowName, types.MergePatchType, patch, metav1.PatchOptions{})
+			if err != nil {
+				isTransientErr := errors.IsTransientErr(err)
```
Comment (alexec) [inline workflow/executor/agent.go:155]: minor - I think think this func logs whether or not the error was transient

**Comment by simster7** [comment]
@sarabala1979 Could I get a review here?

**Comment by alexec** [comment]
@sarabala1979 bump!

**Review by sarabala1979 — APPROVED** [review]

**Review by alexec — COMMENTED** [review]
great stuff!

**Pull request merged**